Templates (#9211)
apple/swift-package-manager: Sources/PackageManagerDocs/Documentation.docc/Templates.md

#  Creating Package Templates

> This guide provides a brief overview of Swift Package Manager templates, describes how a package can make use of templates, and shows how to get started writing your own templates.

Reviewer: ```suggestion
Create a template plugin for Swift Package Manager to generate custom Swift packages based on the custom inputs that you define.
```

^ The first sentence in the article is called an "Abstract" and we generally want it to be a short, imperative sentence that tells the reader what they'll be learning if they jump in and read it. The abstract is presented alongside the title to the create a sort of description of what they'll learn and what core problem its solving.

I suspect there could easily be better title and abstract choices here, but wanted to at least suggest a placeholder that was closer to what we should be aiming for.

## Overview

> Custom _templates_ allow generation of packages whose functionality goes beyond the hard-coded templates provided by Swift Package Manager. Package templates are written in Swift using Swift Argument Parser and the `PackagePlugin` API provided by the Swift Package Manager.

Reviewer: ```suggestion
Custom _templates_ allow Swift Package Manager to generate of packages whose functionality goes beyond the hard-coded templates. Package templates are written in Swift using Swift Argument Parser to provide arguments for generating a Swift package.
```

> A template is represented in the SwiftPM package manifest as a target of the `templateTarget` type and should be available to other packages by declaring a corresponding `template` product. Source code for a template is normally located in a directory under the `Templates` directory in the package, but this can be customized. However, as seen below, authors will also need to write the source code for a plugin.

Reviewer: ```suggestion
Swift Package Manager represents a template in the package manifest as a target of the `templateTarget` type.
Make the template available to other packages by declaring a corresponding `template` product.
Source code for a template is normally located in a directory under the `Templates` directory in the package, but you can customize its location.
The template also requires a template plugin to generate the content for the Swift package.
```

Templates are an abstraction of two types of modules:

> - a template _executable_ that performs the file generation and project setup

Reviewer: ```suggestion
- a template _executable_ that generates the file generation and provides the arguments for the template.
```

The executable allows authors to define user-facing interfaces which gather important consumer input needed by the template to run, using Swift Argument Parser for a rich command-line experience with subcommands, options, and flags.

To learn about using a package template, read <doc:CreatingSwiftPackage#Creating-a-Package-based-on-a-custom-template>.

## Writing a Template

The first step when writing a package template is to decide what kind of template you need and what base package structure it should start with. Templates can build off of any kind of Swift package: executables, libraries, plugins, or even empty packages that will be further customized.

### Declaring a template in the package manifest

> Like all package components, templates are declared in the package manifest. This is done using a `templateTarget` entry in the `targets` section of the package. Templates must be visible to other packages in order to be ran. Thus, there needs to be a corresponding `template` entry in the `products` section as well:

Reviewer: ```suggestion
Like all package components, declare templates in the package manifest. Use the `templateTarget` entry in the `targets` section of the package. Templates must be visible to other packages in order to run. Thus, there needs to be a corresponding `template` entry in the `products` section as well:
```

```swift

> // swift-tools-version: 6.1

Reviewer: ```suggestion
// swift-tools-version: 6.3
```

    ],
    targets: [
        .template(
            name: "LibraryTemplate",
            dependencies: [
                .product(name: "ArgumentParser", package: "swift-argument-parser"),
            ],
            initialPackageType: .library,
            description: "Generate a Swift library package"
        ),
        .template(
            name: "ExecutableTemplate", 
            dependencies: [
                .product(name: "ArgumentParser", package: "swift-argument-parser"),
            ],
            initialPackageType: .executable,
            templatePermissions: [
                .writeToPackageDirectory(reason: "Generate source files and documentation"),
            ],
            description: "Generate an executable package with optional features"
        ),
    ]
)
```

The `templateTarget` declares the name and capability of the template, along with its dependencies. The `initialPackageType` specifies the base package structure that SwiftPM will set up before invoking the template — this can be `.library`, `.executable`, `.tool`, `.buildToolPlugin`, `.commandPlugin`, `.macro`, or `.empty`.

> The Swift script files that implement the logic of the template are expected to be in a directory named the same as the template, located under the `Templates` subdirectory of the package. The template also expects Swift script files in a directory with the same name as the template, alongside a `Plugin` suffix, located under the `Plugins` subdirectory of the package.

Reviewer: ```suggestion
The Swift Package Manager expects the Swift script files that implement the logic of the template to be in a directory with the same as the template, located within a `Templates` subdirectory of the package. The Package Manager also expects Swift script files in a directory with the same name as the template, alongside a `Plugin` suffix, located under the `Plugins` subdirectory of the package.
```

This is complex enough that it's a bit hard to parse right off the bat. I'd suggest that we include an example directory structure inline here - such as the output of the `tree` command, illustrating the expectations to make it easier to understand.

> The `template` product is what makes the template visible to other packages. The name of the template product must match the name of the target.

Reviewer: ```suggestion
Declare the `template` product to make the template visible to other packages. The name of the template product must match the name of the target.
```

#### Template target dependencies

> The dependencies specify the packages that will be available for use by the template executable. Each dependency can be any package product. Commonly this includes Swift Argument Parser for command-line interface handling, but can also include utilities for file generation, string processing, or network requests if needed.

Reviewer: ```suggestion
The dependencies specify the packages that available for use by the template executable. Each dependency can be any package product. A common example is the Swift Argument Parser, which provides arguments, options, and flags for command-line interface handling, but can also include utilities for file generation, string processing, or network requests.
```

    var auth: Bool = false

    func run() throws {
        print("Generating web app with \(framework.rawValue) framework")
        
        if template.documentation {
            print("Including comprehensive documentation")
        }
        
        // Generate web app specific files...
    }
}

struct CLI: ParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "cli",
        abstract: "Generate a command-line tool"
    )

    @ParentCommand var template: MultiVariantTemplate

    @Flag(help: "Include shell completion support")
    var completion: Bool = false

    func run() throws {
        template.run()
        print("Generating CLI tool")
        // Generate CLI specific files...
    }
}

enum WebFramework: String, ExpressibleByArgument, CaseIterable {
    case vapor, hummingbird
}
```

Subcommands can access shared logic and state from their parent command using the `@ParentCommand` property wrapper. This enables a clean seperation of logic between the different layers of commands, while still allowing sequential execution and reuse of common configuration or setup code define at the higher levels.

## Testing Templates

> SwiftPM provides a built-in command for testing templates comprehensively:

Reviewer: ```suggestion
Swift Package Manager provides a built-in command for testing templates:
```
